### PR TITLE
[pipeline](refactor) refactor pipeline task schedule logics

### PR DESCRIPTION
**Pull request by Gabriel39**
## Proposed changes

Issue Number: close #xxx

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: be/src/vec/exec/scan/vscan_node.cpp [hunk 1]
```diff
@@ -323,11 +323,13 @@ Status VScanNode::close(RuntimeState* state) {
...
 void VScanNode::release_resource(RuntimeState* state) {
     if (_scanner_ctx.get()) {
-        if (!state->enable_pipeline_exec() || _should_create_scanner) {
+        if (!state->enable_pipeline_exec()) {
             // stop and wait the scanner scheduler to be done
             // _scanner_ctx may not be created for some short circuit case.
             _scanner_ctx->set_should_stop();
             _scanner_ctx->clear_and_join(this, state);
+        } else if (_should_create_scanner) {
+            _scanner_ctx->clear_and_join(this, state);
```
Comment (wangbo) [inline be/src/vec/exec/scan/vscan_node.cpp:332]: why not set should stop when close?

File: be/src/pipeline/task_scheduler.cpp [hunk 2]
```diff
@@ -328,27 +328,26 @@ void TaskScheduler::_do_work(size_t index) {
...
-            }
         }
-        task->set_state(state);
-        task->set_close_pipeline_time();
-        task->fragment_context()->close_a_pipeline();
+        DCHECK(!task->is_pending_finish()) << task->debug_string();
     }
+    task->set_state(state);
+    task->set_close_pipeline_time();
+    task->fragment_context()->close_a_pipeline();
```
Comment (wangbo) [inline be/src/pipeline/task_scheduler.cpp:350]: Why close a pipeline here? even task->is_pending_finish()?

File: be/src/vec/exec/scan/vscan_node.cpp (hunk 1 above)
Comment (Gabriel39) [reply be/src/vec/exec/scan/vscan_node.cpp:332]: because it was already set in `try_close`

Comment (Gabriel39) [reply be/src/vec/exec/scan/vscan_node.cpp:332]: because it is already set in `try close`

**Comment by Gabriel39** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 45.26 seconds
 stream load tsv:          521 seconds loaded 74807831229 Bytes, about 136 MB/s
 stream load json:         19 seconds loaded 2358488459 Bytes, about 118 MB/s
 stream load orc:          65 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          30 seconds loaded 861443392 Bytes, about 27 MB/s
 insert into select:          29.6 seconds inserted 10000000 Rows, about 337K ops/s
 storage size: 17169800794 Bytes

**Review by wangbo — DISMISSED** [review]
LGTM

File: be/src/exec/data_sink.h [hunk 3]
```diff
@@ -67,7 +67,7 @@ class DataSink {
         return Status::NotSupported("Not support send block");
     }
 
-    virtual void try_close(RuntimeState* state, Status exec_status) {}
+    virtual Status try_close(RuntimeState* state, Status exec_status) { return Status::OK(); }
```
Comment (yiguolei) [inline be/src/exec/data_sink.h:70]: add [[nodiscard]]  attribute

**Comment by Gabriel39** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 45.38 seconds
 stream load tsv:          506 seconds loaded 74807831229 Bytes, about 140 MB/s
 stream load json:         19 seconds loaded 2358488459 Bytes, about 118 MB/s
 stream load orc:          64 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          31 seconds loaded 861443392 Bytes, about 26 MB/s
 insert into select:          29.7 seconds inserted 10000000 Rows, about 336K ops/s
 storage size: 17169212248 Bytes

**Comment by Gabriel39** [comment]
run buildall

**Review by HappenLee — DISMISSED** [review]
LGTM

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 46.17 seconds
 stream load tsv:          510 seconds loaded 74807831229 Bytes, about 139 MB/s
 stream load json:         20 seconds loaded 2358488459 Bytes, about 112 MB/s
 stream load orc:          65 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          31 seconds loaded 861443392 Bytes, about 26 MB/s
 insert into select:          29.9 seconds inserted 10000000 Rows, about 334K ops/s
 storage size: 17162497475 Bytes

File: be/src/vec/sink/vtablet_sink.cpp [hunk 4]
```diff
@@ -1320,11 +1320,7 @@ void VOlapTableSink::_cancel_all_channel(Status status) {
             print_id(_load_id), _txn_id, status);
 }
 
-void VOlapTableSink::try_close(RuntimeState* state, Status exec_status) {
-    if (_try_close) {
```
Comment (yiguolei) [inline be/src/vec/sink/vtablet_sink.cpp:1324]: why remove this line?

**Comment by Gabriel39** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 45.25 seconds
 stream load tsv:          541 seconds loaded 74807831229 Bytes, about 131 MB/s
 stream load json:         19 seconds loaded 2358488459 Bytes, about 118 MB/s
 stream load orc:          64 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          30 seconds loaded 861443392 Bytes, about 27 MB/s
 insert into select:          29.4 seconds inserted 10000000 Rows, about 340K ops/s
 storage size: 17167819717 Bytes

**Comment by Gabriel39** [comment]
run buildall

**Comment by Gabriel39** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 45.15 seconds
 stream load tsv:          505 seconds loaded 74807831229 Bytes, about 141 MB/s
 stream load json:         20 seconds loaded 2358488459 Bytes, about 112 MB/s
 stream load orc:          64 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          31 seconds loaded 861443392 Bytes, about 26 MB/s
 insert into select:          29.9 seconds inserted 10000000 Rows, about 334K ops/s
 storage size: 17162775042 Bytes

**Review by wangbo — APPROVED** [review]
LGTM

**Review by yiguolei — APPROVED** [review]
LGTM

**Pull request merged**